crates/sui-indexer-alt/src/pipeline/sequential/committer.rs

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cmp::Ordering, collections::BTreeMap, sync::Arc};
+
+use diesel_async::{scoped_futures::ScopedFutureExt, AsyncConnection};
+use mysten_metrics::spawn_monitored_task;
+use tokio::{
+    sync::mpsc,
+    task::JoinHandle,
+    time::{interval, MissedTickBehavior},
+};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, info, warn};
+
+use crate::{
+    db::Db,
+    metrics::IndexerMetrics,
+    models::watermarks::CommitterWatermark,
+    pipeline::{Indexed, PipelineConfig, LOUD_WATERMARK_UPDATE_INTERVAL, WARN_PENDING_WATERMARKS},
+};
+
+use super::Handler;
+
+/// The committer task gathers rows into batches and writes them to the database.
+///
+/// Data arrives out of order, grouped by checkpoint, on `rx`. The task orders them and waits to
+/// write them until either a configural polling interval has passed (controlled by
+/// `config.collect_interval`), or `H::BATCH_SIZE` rows have been accumulated and we have received
+/// the next expected checkpoint.
```

Comment:
In a sense then, the sequential committer is the combined concurrent collector, committer, and watermark task

Reply:
In fact, the committer implementation here looks a lot like the watermark task. So iiuc, in the concurrent setting, the collector, committer, and watermark tasks are largely independent, and can batch, write, and update watermarks at its own cadence. Not the same for the sequential setting, understandably

Reply:
That's right -- initially the concurrent pipeline was similar, but when I implemented concurrent writes, I needed to split it up.